org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java

```diff
@@ -0,0 +1,185 @@
+
+		StringBuilder insertString = new StringBuilder();
+		if (docType.getInternalSubset() == null) {
+			String doctypeContent = docType.getTextContent();
+			// Add subset
+			if (Character.isWhitespace(doctypeContent.charAt(doctypeContent.length() - 2))) {
+				insertString.append("[");
+			} else {
+				insertString.append(" [");
+			}
+		}
+
+		if (insertPosition.getCharacter() != 0) {
+			insertString.append(delimiter);
+		}
+
+		insertString.append(createEntityDeclaration(entityName, settings));
+
+		if (docType.getInternalSubset() == null) {
+			insertString.append(delimiter + "]");
+		} else if (docTypeEnd.getLine() == insertPosition.getLine()) {
+			insertString.append(delimiter);
+		}
+
+		CodeAction action = CodeActionFactory.insert(message, insertPosition, insertString.toString(),
+				document.getTextDocument(), diagnostic);
+		codeActions.add(action);
+	}
+
+	private Position getEntityInsertPosition(DOMDocument document) throws BadLocationException {
+		TextDocument textDocument = document.getTextDocument();
+		DOMDocumentType docType = document.getDoctype();
+
+		String subset = docType.getInternalSubset();
+		if (subset == null) { // subset ([]) does not exist
+			return textDocument.positionAt(docType.getEnd() - 1);
+		}
+
+		DOMNode lastChild = docType.getLastChild();
+		if (lastChild != null) {
+			return textDocument.positionAt(lastChild.getEnd());
+		}
+
+		// empty subset exists
+		String subsetValue = "[" + subset + "]";
+		int index = docType.getTextContent().indexOf(subsetValue);
+		if (index >= 0) {
+			// if statement should always satisfy
+			// index of ]
+			index += subsetValue.length() - 1;
+			return docType.getOwnerDocument().positionAt(index + docType.getStart());
+		}
+
+		return null;
+	}
+
+	private void addDoctypeAndEntityCodeAction(String entityName, Diagnostic diagnostic, DOMDocument document,
+			SharedSettings settings, List<CodeAction> codeActions) throws BadLocationException {
+		Position insertPosition;
+		if (document.hasProlog()) {
+			insertPosition = new Position(1, 0);
+		} else {
+			insertPosition = new Position(0, 0);
+		}
+
+		String delimiter = document.lineDelimiter(insertPosition.getLine());
+		String message = "Declare DOCTYPE containing ENTITY " + entityName;
+		StringBuilder insertString = new StringBuilder();
+		insertString.append("<!DOCTYPE article [" + delimiter);
+		insertString.append(createEntityDeclaration(entityName, settings));
+		insertString.append(delimiter);
+		insertString.append("]>" + delimiter);
+		CodeAction action = CodeActionFactory.insert(message, insertPosition, insertString.toString(),
+				document.getTextDocument(), diagnostic);
+		codeActions.add(action);
+	}
+
+	private String getEntityName(Diagnostic diagnostic, Range range, DOMDocument doc) throws BadLocationException {
+		String name = doc.getText().substring(doc.offsetAt(range.getStart()), doc.offsetAt(range.getEnd()));
+		String removedAmpAndSemiColon = name.substring(1, name.length() - 1);
```

Comment:
Here, I am retrieving the entity name from the error range, and then checking it against the error message.
It would be better if we could store the name in the `Diagnostic` when the `Diagnostic` is being created.

Related issue: https://github.com/eclipse/lemminx/issues/681

Reply:
Indeed it's little ugly but for the moment we have not teh choice to do that. Please add a comments for that and set the link to the LSP issue (and please do that for the code #681)